src/main/scala/lib/HList.scala

```diff
@@ -1,87 +1,30 @@
-  trait LowPrioMapper extends LowestPrioMapper {
-    implicit def iteratedMapper[S, H, T <: HList, B, tOut <: HList](implicit iTail: Mapper[S, T, B, tOut]): Mapper[S, H :: T, B, H :: tOut] =
-      new Mapper[S, H :: T, B, H :: tOut] {
-        def apply(hc: H :: T, f: S => B) = hc.head :: iTail(hc.tail, f)
-      }
+  sealed trait KMapper[S[_], HL <: HList, B[_], Out <: HList] {
+    def apply(hl: HL, f: S ~> B): Out
   }
 
-  trait LowestPrioMapper {
-    implicit def tipNotFound[S, HC <: HNil, B]: Mapper[S, HC, B, HNil.type] =
-      new Mapper[S, HC, B, HNil.type] {
-        def apply(hc: HC, f: S => B) = HNil
+  object KMapper extends LowPrioMapper {
+    implicit def iteratedMapper[S[_], H, T <: HList, B[_], O, tOut <: HList](implicit iTail: KMapper[S, T, B, tOut]): KMapper[S, S[O] :: T, B, B[O] :: tOut] =
+      new KMapper[S, S[O] :: T, B, B[O] :: tOut] {
+        def apply(hc: S[O] :: T, f: S ~> B) = f(hc.head) :: iTail(hc.tail, f)
       }
   }
 
-  def map[S, B, T <: HList, tOut <: HList](hc: T, f: S => B)(implicit ev: Mapper[S, T, B, tOut]): tOut = ev(hc, f)
-  // --
-
-  sealed trait Extractor[S, HL <: HList] {
-    def apply(hl: HL): S
-  }
-  object Extractor extends LowPrioExtractor {
-    implicit def typeMatchedExtractor[S, H, T <: HList](implicit ev: H =:= S): Extractor[S, H :: T] =
-      new Extractor[S, H :: T] {
-        def apply(hc: H :: T) = ev(hc.head)
+  trait LowPrioMapper {
+    implicit def tipNotFound[S[_], HC <: HNil, B[_]]: KMapper[S, HC, B, HNil.type] =
+      new KMapper[S, HC, B, HNil.type] {
+        def apply(hc: HC, f: S ~> B) = HNil
       }
   }
-  trait LowPrioExtractor {
-    implicit def iteratedExtractor[S, H, T <: HList](implicit iTail: Extractor[S, T]): Extractor[S, H :: T] =
-      new Extractor[S, H :: T] {
-        def apply(hc: H :: T) = iTail(hc.tail)
-      }
-  }
-  def extract[S, T <: HList](hc: T)(implicit ev: Extractor[S, T]) = ev(hc)
-  // --
-  sealed trait ApplyUpdate1[S, HL <: HList, B] {
-    def apply(hl: HL, f: S => (S, B)): (HL, B)
-  }
 
-  object ApplyUpdate1 extends LowPrioApplyUpdate1 {
-    implicit def typeMatchedApplyUpdate1[S, H, T <: HList, B](implicit evSH: S =:= H, evHS: H =:= S): ApplyUpdate1[S, H :: T, B] =
-      new ApplyUpdate1[S, H :: T, B] {
-        def apply(hc: H :: T, f: S => (S, B)) = {
-          val r = f(evHS(hc.head))
-          (evSH(r._1) :: hc.tail, r._2)
-        }
-      }
-  }
-  trait LowPrioApplyUpdate1 {
-    implicit def iteratedApplyUpdate1[S, H, T <: HList, B](implicit iTail: ApplyUpdate1[S, T, B]): ApplyUpdate1[S, H :: T, B] =
-      new ApplyUpdate1[S, H :: T, B] {
-        def apply(hc: H :: T, f: S => (S, B)) = {
-          val r = iTail(hc.tail, f)
-          (hc.head :: r._1, r._2)
-        }
-      }
-  }
-  def applyUpdate1[S, B, T <: HList](hc: T, f: S => (S, B))(implicit ev: ApplyUpdate1[S, T, B]) = ev(hc, f)
+  def kMap[S[_], B[_], T <: HList, tOut <: HList](hc: T, f: S ~> B)(implicit ev: KMapper[S, T, B, tOut]): tOut = ev(hc, f)
```

Comment:
I've tried adding Aux in the same way however that results in the same error.. have I missed something? In general reasoning Aux is kind of similar as the default shapeless mapper...

```
Error:(24, 180) type mismatch;
 found   : Cqrs.BatchRunner[Db,shapeless.::[Cqrs.Projection[D],m.Out]]
 required: Cqrs.BatchRunner[Db,shapeless.::[Cqrs.Projection[D],PROJS]]
Error occurred in an application involving default arguments.
  def addProjection[D](proj: Projection[D])(implicit m: KMapper[Projection, PROJS, Projection]): BatchRunner[Db, Projection[D] :: PROJS] = copy(projections = proj :: projections).runProjections
                                                                                                                                                                                   ^
```